kahypar/utils/validate.h

```diff
@@ -0,0 +1,312 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ ******************************************************************************/
+
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "kahypar/definitions.h"
+#include "kahypar/datastructure/fast_reset_flag_array.h"
+
+namespace kahypar {
+namespace validate {
+
+using ds::FastResetFlagArray;
+
+/*!
+  * Helper class for reading a line and reporting syntactical errors
+  */
+class CheckedIStream {
+ public:
+  explicit CheckedIStream(const char* begin, const char* end, size_t line_number = 0):
+    _current(begin), _next(nullptr), _end(end), _line_number(line_number) {
+      // whitespace at end of line is not considered an error
+      while (_end > _current && *(_end - 1) == ' ') {
+        --_end;
+      }
+    }
+
+  explicit CheckedIStream(const std::string& line, size_t line_number = 0):
+    CheckedIStream(line.c_str(), line.c_str() + line.size(), line_number) { }
+
+  template<typename ResultT>
+  bool operator>>(ResultT& result) {
+    unsigned long long val = std::strtoull(_current, &_next, 10);
+    if (val == 0 && _next == _current && _next != _end) {
+      std::cerr << "Error: Expected positive number";
+      if (_line_number != 0) {
+        std::cerr << " (line " << _line_number << ")";
+      }
+      std::cerr << std::endl;
+      exit(1);
+    }
+
+    const ResultT max = std::numeric_limits<ResultT>::max();
+    if (val > static_cast<unsigned long long>(max)) {
+      std::cerr << "Error: ID is out of range: " << val << ", but maximum is " << max;
+      if (_line_number != 0) {
+        std::cerr << " (line " << _line_number << ")";
+      }
+      std::cerr << std::endl;
+      exit(1);
+    }
+
+    if (_current != _next) {
+      _current = _next;
+      result = static_cast<ResultT>(val);
+      return true;
+    }
+    return false;
+  }
+
+  bool empty() const {
+    char* p = nullptr;
+    std::strtoull(_current, &p, 10);
+    return p == _current && p == _end;
+  }
+
+ private:
+  const char* _current;
+  char* _next;
+  const char* _end;
+  size_t _line_number;
+};
+
+/*!
+  * Types of semantic errors in hypergraph input
+  */
+enum class InputErrorType : uint8_t {
+  FirstHyperedgeIndexNotZero,   // first entry in index array not zero
+  HyperedgeIndexDescending,     // index array is not in ascending order
+  HyperedgeOutOfBounds,         // size of hyperedge (according to index array) is larger than number of nodes
+  HyperedgeSizeZeroOrOne,       // size of hyperedge is zero or one
```

Comment:
While single-pin hyperedges are essentially meaningless for partitioning, I wouldn't treat them as an error.

Reply:
Ah.. just reading a few lines below... you propose to treat them as "not fatal". That makes sense.

Reply:
Exactly. I think it makes sense that a warning is printed for single pin hyperedges

Reply:
Agreed.